ctru-rs/src/applets/swkbd.rs

```diff
@@ -6,26 +6,25 @@
 use crate::services::{apt::Apt, gfx::Gfx};
 use ctru_sys::{
     aptLaunchLibraryApplet, aptSetMessageCallback, envGetAptAppId, svcCloseHandle,
-    svcCreateMemoryBlock, APT_SendParameter, SwkbdButton, SwkbdDictWord, SwkbdExtra,
-    SwkbdLearningData, SwkbdState, SwkbdStatusData, APPID_SOFTWARE_KEYBOARD, APTCMD_MESSAGE,
-    NS_APPID,
+    svcCreateMemoryBlock, APT_SendParameter, SwkbdButton, SwkbdDictWord, SwkbdLearningData,
+    SwkbdState, SwkbdStatusData, APPID_SOFTWARE_KEYBOARD, APTCMD_MESSAGE, NS_APPID,
 };
 
 use bitflags::bitflags;
 
+use std::borrow::Cow;
 use std::ffi::{CStr, CString};
 use std::fmt::Display;
 use std::iter::once;
 use std::str;
 
-type CallbackFunction = dyn Fn(&CStr) -> (CallbackResult, Option<CString>);
+type CallbackFunction = dyn Fn(&str) -> (CallbackResult, Option<Cow<'static, str>>);
```

Comment:
Shouldn't this be `Cow<'a>` where `'a` is the lifetime of the input string? i.e.
```suggestion
type CallbackFunction<'a> = dyn Fn(&'a str) -> (CallbackResult, Option<Cow<'a, str>>);
```

Reply:
It doesn't matter because the thing we actually work with in the API is `Option<Box<CallbackFunction>>`, and that has an implicit `'static` bound.

Reply:
Hm, is there really any reason to make it a `Cow` then? I thought the idea with that was we could reuse the incoming string's lifetime, but it seems like with the `'static` bound might as well just make it a `String`

Reply:
Take a look at [SoftwareKeyboard::set_filter_callback](https://rust3ds.github.io/ctru-rs/crates/ctru/applets/swkbd/struct.SoftwareKeyboard.html#method.set_filter_callback) to see how this function is actually used. The input string contains text that was just typed into the Software Keyboard. The output is a user-supplied response message that the keyboard can reply with based on whatever logic you have written.

Notice how in the current API, the user is forced to heap-allocate their response message. The purpose of using `Cow` is so that the user doesn't have to do that unless they want to reply with a formatted message.

Reply:
ahhh, ok, so the idea would be so you could return `Cow::from("foo")` or something, but not necessarily `Cow::from(input_str)` (would need `Cow::from(input_str.to_string()`?). Is that understanding accurate?

Reply:
Yep, you got it.